src/Style.hs

```diff
@@ -283,12 +269,40 @@ attribute = identifier -- TODO: Naming convention - same as identifiers?
 -- styPrettyPrint b = ""
 
 --------------------------------------------------------------------------------
--- Functions used by "Runtime"
+-- Functions used by "Runtime" -- Style Translator
+-- Translates from a Style AST to:
+-- * a list of geometric primitives
+-- * a list of constraints
+-- * a list of objectives
+
+-- Type aliases for readability in this section
+-- | 'StyContext' maintains the current output of the translater
+type StyContext a =
+    (StyDict,         -- | dictionary mapping Substance ID to Style objects
+    [ObjFnInfo a],    -- | List of objective functions
+    [ConstrFnInfo a]) -- | List of constraints
+
+-- | A dictionary storing properties of a Style object, e.g. "start" for 'Arrow'
+type Properties = M.Map String Expr
+
+-- | A Style object is a type annotation ('StyType') and a collection of
+-- properties. E.g. an arrow is 'Arrow' and properties: "start", "end".
+type StyObj = (StyType, Properties)
+
+-- | a Style Dictionary maps from a Substance id to a 'StySpec' -- collection of Style objects and some metadata about the Substance object
+type StyDict = M.Map Name StySpec
 
------ Parser for Style design
+-- | Style specification for a particular object declared in Substance (declarations and constraints)
+-- (TODO: maybe this is not the best model, since there is no difference between "primary" and "secondary" shapes)
+-- NOTE: for Substance constraints such as `Subset A B`, the 'spId' will be '_Subset_A_B' and the 'spArgs' will be '['A', 'B']'
+data StySpec = StySpec {
+    spType   :: C.SubType,  -- | The Substance type of the object
+    spId     :: String,     -- | The Substance ID of the object
+    spArgs   :: [String],   -- | the "arguments" following the type.  The idea is to capture @f@, @A@ and @B@ in the case of @Map f A B@, which is needed for pattern matching (TODO: Maybe not the best term here.)
```

Comment:
do we ever use spArgs? what is it used for?

Reply:
> "The idea is to capture @f@, @A@ and @B@ in the case of @Map f A B@, which is needed for pattern matching (TODO: Maybe not the best term here.)"

So this `spArgs` field exists because some Substance objs are defined by an n-tuple. For instance, `Subset A B` is effectively `(A, B)` with a type `Subset`. Therefore, to pattern match on them, we need a record of all arguments in a tuple.